src/ledctl.c

```diff
@@ -668,8 +534,6 @@ static ledctl_status_code_t _cmdline_ibpi_parse(int argc, char *argv[])
 			}
 		}
 	}
-	if (_ibpi_state_determine(&ibpi_list) != LEDCTL_STATUS_SUCCESS)
```

Comment:
I've been testing `ledctl` with these changes and it seems to work fine, but I could certainly be missing something.  I spent quite a bit of time looking at this code before I took it out.  I'll review again.  If you can point out a command line example that would show the breakage, that would be great to help with my understanding.

Reply:
OK, I believe I removed this as I was trying to reduce coupling between ledctl and the library.  This determine code is accessing the block structure which isn't visible in the library.  That being said I still don't understand why this is needed.


In the function `_determine`
```c
static void _determine(struct ibpi_state *state)
{
	if (list_is_empty(&state->block_list) == 0) {
		struct block_device *block;

		list_for_each(&state->block_list, block) {
			if (block->ibpi != state->ibpi)
				block->ibpi = state->ibpi;
		}
	} else {
		log_warning
		    ("IBPI %s: missing block device(s)... pattern ignored.",
		     ibpi2str(state->ibpi));
	}
}
```
We walk the list of the ibpi values the user requested on the command line `state->ibpi` and compare it to the value from the `struct block_device`, via `block->ibpi`.  If they are different we make them equal by changing the block value.  Then when we call `_ledctl_execute`

```c
tatic ledctl_status_code_t _ledctl_execute(struct list *ibpi_local_list)
{
	struct ibpi_state *state;
	struct block_device *device;

	if (!listed_only) {
		list_for_each(sysfs_get_block_devices(), device)
			device->send_fn(device, IBPI_PATTERN_LOCATE_OFF);
	}

	list_for_each(ibpi_local_list, state)
		list_for_each(&state->block_list, device) {
			if (state->ibpi != device->ibpi) {
				log_debug("Mismatch detected for %s, ibpi state: %s, device state %s\n",
					  device->sysfs_path, state->ibpi,
					  device->ibpi);
				return LEDCTL_STATUS_IBPI_DETERMINE_ERROR;
			}
			device->send_fn(device, device->ibpi);
		}

	list_for_each(sysfs_get_block_devices(), device)
		device->flush_fn(device);

	return LEDCTL_STATUS_SUCCESS;
}
```

We compare them again, and if they are different we log a warning and return an error.  My questions are:

1. How can they ever be different as we make sure they are the same before we get to this code?
2. Why do we need to do this compare instead of just taking what the user provided on the command line and passing it as an argument to the `send_fn`?  User wants this device to have this IPBI value, so send it on to the hardware.

The change I made does 2 indirectly by calling the library function to set the LED by device node and IPBI value.

I'm hoping that when you answer questions 1 & 2 above, that I'll understand what I'm missing.

Reply:
Hi,
Sorry for late response.  Hope, It will clarify your objections.

> 1. How can they ever be different as we make sure they are the same before we get to this code?

The verification in `_ledctl_execute` is dead code, totally agree. Probably, my motivation of this warning was to have the verification in case when someone will change the logic.  The warning gives reader prove that block devices added to ibpi_state list should have same state under ibpi. It is like self-documentation.
Anyway, I'm fine with removing it as it is dead code.

> 2. Why do we need to do this compare instead of just taking what the user provided on the command line and passing it as an argument to the `send_fn`?  User wants this device to have this IPBI value, so send it on to the hardware.

See above. 

> The change I made does 2 indirectly by calling the library function to set the LED by device node and IPBI value.
> 
> I'm hoping that when you answer questions 1 & 2 above, that I'll understand what I'm missing.

That is good change, but we need to dig into implementation details around `_ibpi_state_determine`.
First, see this commit: https://github.com/intel/ledmon/commit/e027ad6c488e799b7fe242438b6a70c293a0f1a4
Previously we used `device->state` in  `_ledctl_execute`:
```C
device->send_fn(device, device->ibpi);
```
And you changed that:
```C
led_set(ctx, device, state->ibpi);
```
That should be fine. We could have different state under `device->state` but it shouldn't matter now. It also explains why my warning is no longer necessary. But you also removed `_ibpi_state_determine`. It is a part of parsing method, see the error at the end of the function, I'm expecting that result of negative scenarios is different now.
We need to take care of user interface, I would like to avoid the regressions here. 

Please also test negative scenarios, for example:
```
#ledctl locate={} normal=/dev/nvme1n1
#ledctl normal=
```
I expect differences there.